### PR TITLE
Implement Exception Sampling in the Python SDK

### PR DESCRIPTION
**Pull request by rohdesamuel**
This utilizes the ExecutionContext to set the current OutputSampler for the the input value.

Addresses #26704

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by rohdesamuel** [comment]
R: @udim

File: sdks/python/apache_beam/runners/worker/bundle_processor.py [hunk 1]
```diff
@@ -1181,10 +1182,17 @@ def shutdown(self):
...
 
-class ExecutionContext(object):
-  def __init__(self):
-    self.delayed_applications = [
-    ]  # type: List[Tuple[operations.DoOperation, common.SplitResultResidual]]
+@dataclass
+class ExecutionContext:
+  # Any splits to be processed later.
+  delayed_applications = [
+  ]  # type: List[Tuple[operations.DoOperation, common.SplitResultResidual]]
```
Comment (rohdesamuel) [inline sdks/python/apache_beam/runners/worker/bundle_processor.py:1189]: TODO: change this to a py3 type annotation

File: sdks/python/apache_beam/runners/worker/operations.py [hunk 2]
```diff
@@ -178,7 +177,10 @@ def __init__(self,
     self.output_index = output_index
     self.coder = coder
     self.consumers = consumers
-    self.element_sampler = element_sampler
+    self.exception_sampler = output_sampler
+    self.element_sampler = (
+        output_sampler.element_sampler if output_sampler else None)
+    self.execution_context = None  # type: Optional[ExecutionContext]
```
Comment (rohdesamuel) [inline sdks/python/apache_beam/runners/worker/operations.py:183]: TODO: change this type to annotation

File: sdks/python/apache_beam/runners/worker/operations.py [hunk 3]
```diff
@@ -178,7 +177,10 @@ def __init__(self,
     self.output_index = output_index
     self.coder = coder
     self.consumers = consumers
-    self.element_sampler = element_sampler
+    self.exception_sampler = output_sampler
```
Comment (rohdesamuel) [inline sdks/python/apache_beam/runners/worker/operations.py:180]: TODO: rename exception_sampler to output_sampler

File: sdks/python/apache_beam/runners/common.py [hunk 4]
```diff
@@ -1417,9 +1422,26 @@ def process(self, windowed_value):
     try:
       return self.do_fn_invoker.invoke_process(windowed_value)
     except BaseException as exn:
+      self._maybe_sample_exception(exn, windowed_value)
       self._reraise_augmented(exn)
       return []
 
+  def _maybe_sample_exception(
+      self, exn: BaseException, windowed_value: Any) -> None:
```
Comment (rohdesamuel) [inline sdks/python/apache_beam/runners/common.py:1430]: TODO: Change windowed_value type to WindowedValue[Any] (or something similar)

File: sdks/python/apache_beam/runners/common.py [hunk 5]
```diff
@@ -1417,9 +1422,26 @@ def process(self, windowed_value):
...
+    if self.execution_context is None:
+      return
+
+    exception_sampler = self.execution_context.exception_sampler
+    if exception_sampler is None:
+      return
+
+    exception_sampler.sample_exception(
+        windowed_value,
+        exn,
```
Comment (rohdesamuel) [inline sdks/python/apache_beam/runners/common.py:1441]: TODO: Get the exception with stacktrace

File: sdks/python/apache_beam/runners/worker/data_sampler.py [hunk 6]
```diff
@@ -96,6 +110,7 @@ def __init__(
     self._sample_timer = SampleTimer(sample_every_sec, self)
     self.element_sampler = ElementSampler()
     self.element_sampler.has_element = False
+    self._exceptions: Deque[Any] = collections.deque(maxlen=max_samples)
```
Comment (rohdesamuel) [inline sdks/python/apache_beam/runners/worker/data_sampler.py:113]: Type should be Tuple[Any, ExceptionMetadata]

File: sdks/python/apache_beam/runners/worker/data_sampler.py [hunk 7]
```diff
@@ -115,19 +130,47 @@ def remove_windowed_value(self, el: Union[WindowedValue, Any]) -> Any:
...
+        ]
+        samples = [
+            self.remove_windowed_value(s) for s in self._samples
+            if s not in seen
+        ]
 
       # Encode in the nested context b/c this ensures that the SDK can decode
       # the bytes with the ToStringFn.
       if clear:
         self._samples.clear()
```
Comment (rohdesamuel) [inline sdks/python/apache_beam/runners/worker/data_sampler.py:156]: TODO: clear _exceptions

File: sdks/python/apache_beam/runners/worker/data_sampler.py [hunk 8]
```diff
@@ -136,6 +179,13 @@ def sample(self) -> None:
         self.element_sampler.has_element = False
         self._samples.append(self.element_sampler.el)
 
+  def sample_exception(
+      self, el: Any, exn: BaseException, transform_id: str,
+      instruction_id: str) -> None:
+    """Adds the given exception to the samples."""
+    self._exceptions.append(
```
Comment (rohdesamuel) [inline sdks/python/apache_beam/runners/worker/data_sampler.py:186]: TODO: use the _samples_lock here

**Review by udim — APPROVED** [review]

**Pull request merged**